### DISCUSSION TITLE
controlling when menu appears? (some other questions)

### DISCUSSION THREAD

**Discussion by emaayan** [davetcc/tcMenu#328] [Q&A] [answered]
hi.. been starting to use tcMenu it looks nice. but i have some questions. 

is there a way to have the the menu appear only by pressing a button (or actualy long pressing a button) 
addtionally can i use the same button i've  configured  for ok in the menu, or do i need a dedicated button ? 
i'm trying to create a clock and the menu is intentended to configure the date and time. 
so the default operation should be when you switch it on it would automatically show the time 
or if if it's a new card, only then show the menu. 

another question is using font, i need to use hebrew fonts for the clock but i don't mind using the other fonts for the menu, is there a way to keep switching between the two? 

i tried using the date config menu option, for some reason it allows have a month as 0 
is there a way to do validation on date input for example checking that day number does not exceed the total number of days in a month, and if it does reset it back to the maximum date

**Comment by davetcc** [comment]
See taking over the display, it works the opposite way around, but does exactly what you want, you take over the display, and give it back at some point later, you're provided with if the select button has been pressed, and the current encoder value. In fact it is such a popular option it has it's own page::

https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/renderer-take-over-display/

You don't mention which displays you're using but to a large extent you can use a few fonts at once as you want with most libraries. Even within the menu you can have some menus using different display properties. Many of the examples cover this usage. If you use TcUnicode, it fully supports UTF8 on most libraries (including Adafruit).

Sounds like there could be a bug in the date parser if you can confirm that please raise a bug, but you can also now override the date renderFN using designer UI and provide your own stricter checking.

https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/menu-item-types/based-on-runtimemenuitem/

**Reply by emaayan** [reply]
oh i'm using 2.42 oled displayh from diymore 
you don't have that display so i had to use the custom option 
#define CS 5
#define DC 17   // 16
#define RESET 4 // 17
U8G2_SSD1309_128X64_NONAME0_F_4W_HW_SPI _oled(U8G2_R1, CS, DC, RESET);

i wasn't sure what other code to give it other than .begin()

**Reply by emaayan** [reply, marked as answer]
> See taking over the display, it works the opposite way around, but does exactly what you want, you take over the display, and give it back at some point later, you're provided with if the select button has been pressed, and the current encoder value. In fact it is such a popular option it has it's own page::
> 
> https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/renderer-take-over-display/

so you're saying i should cal
renderer.takeOverDisplay(myDisplayCallback);
in in the callback have all my code that draws the display include calling the RTC DS3231 api's , as long as i don't see the value for the select button being pressed  RPRESS_HELD, when i do see it i need to call renderer.giveBackDisplay(), from inside the callback 
right? 
like so...
```
void myDisplayCallback(unsigned int encoderValue, RenderPressMode clicked) {    
    if (clicked==RPRESS_HELD){
        renderer.giveBackDisplay();
    }
}

void setup() {
    Serial.begin(115200);
    _oled.begin();
    //_oled.setFont(u8g2_font_6x13_t_hebrew);
    setupMenu();   
    renderer.takeOverDisplay(myDisplayCallback);
}

```

**Comment by davetcc** [comment]
That sounds about right, yes, for the custom U8G2 option you just create the variable and initialize it before setupMenu() is called.

**Reply by emaayan** [reply]
you might want to add that for the custom wizard, cause initially i just used the variable and expected to init by the framework. 
how do i verify the date bug?

**Comment by davetcc** [comment]
> you might want to add that for the custom wizard

We can't now, there are at least 100s if not 1000s of menus relying on that working how it does.

> how do i verify the date bug?

I don't think that 0 should be allowed in the month or day field, if that is allowed, it is a bug. Even the days in the month should be correct for all gregorian calendars. If you have cases that are clearly wrong, please raise an issue and fully document how to recreate.

**Reply by emaayan** [reply]
oh i meant just add it as text to the wizard.. 
cause when the code is generated i'm not sure what api's i can or can't call
like can i call begin, setFront, can i call clearBuffer, DO i need to call clearBuffer, etc..
i'm not very familiar with the library so those are the question i asked my self when it was done. 

![image](https://user-images.githubusercontent.com/13434544/231735484-0f3ac484-dabf-49f0-98f2-b15871e9ea5b.png)

so this would be a bug right? (does max length have any effect on this? cause mine is 5) 

![image](https://user-images.githubusercontent.com/13434544/231735797-3ffe4ef1-07e2-4225-b660-523aac4c8048.png)

**Reply by davetcc** [reply]
I'll try and make the description a bit clearer and check the associated online help page (link - "click for documentation")
Yes, that does not look like a good date, please raise as an issue.

**Reply by emaayan** [reply]
here you go..hope this helps..

https://github.com/davetcc/tcMenu/issues/332
